File: Two_points/670.maximum-swap.go
package leetcode

import "strconv"

// @leet start
func maximumSwap(num int) int {
	input := []rune(strconv.Itoa(num))
	numIdx := make([]int, 10)
	// record the last freq of each integer
	for i, r := range input {
		numIdx[int(r-'0')] = i
	}

	for i, r := range input {
		// if an large integer show in later
		for v := 9; v > int(r-'0'); v-- {
			if laterIdx := numIdx[v]; laterIdx > i {
				input[laterIdx], input[i] = input[i], input[laterIdx]
				res, _ := strconv.Atoi(string(input))
				return res
			}
		}
	}
	return num
}

// @leet end

File: TwoPointer/670.maximum-swap.py
# @leet start
class Solution:
    def maximumSwap(self, num: int) -> int:
        bestLoc = [0] * 10
        for i, v in enumerate(str(num)):
            bestLoc[int(v)] = i
        #
        numList = list(str(num))
        #
        for i, curInt in enumerate(numList):
            for v in range(9, int(curInt), -1):
                if bestLoc[v] > i:
                    numList[i], numList[bestLoc[v]] = numList[bestLoc[v]], numList[i]
                    return int("".join(numList))
        return num


# @leet end

